addon: exclude virtual modules from Vite's optimizeDeps pre-bundling

A consumer reported `✘ [ERROR] Could not resolve "virtual:swatchbook/tokens"` running `storybook dev` (while `storybook build` worked). Root cause: the addon's virtual-module plugin has no `config` hook excluding its virtual IDs from Vite's `optimizeDeps` pre-bundling.

Vite uses **esbuild** for `optimizeDeps` pre-bundling (development mode). esbuild doesn't see Rollup-style `resolveId` hooks — only Rollup / Vite's full plugin pipeline does. When Vite's dep-bundling step scans a preview file that imports `virtual:swatchbook/tokens` (e.g., from `@unpunnyfuns/swatchbook-addon/hooks`'s `use-token.ts`), esbuild has no resolver registered for that ID and the error fires.

Build mode uses Rollup all the way through, no esbuild pre-bundle, which is why `storybook build` succeeds. Latent issue triggered by particular consumer setups (story or preview imports that get pulled into pre-bundling).

## Fix

Add a `config` hook to the virtual plugin that registers our virtual IDs with `optimizeDeps.exclude`:

```ts
config() {
  return {
    optimizeDeps: {
      exclude: [VIRTUAL_MODULE_ID, INTEGRATION_SIDE_EFFECTS_VIRTUAL_ID],
    },
  };
}
```

Tells Vite "don't pre-bundle these; route through the plugin pipeline." Canonical pattern for any Vite plugin owning a virtual module.

## Test

Assert the plugin's `config` hook returns the expected `optimizeDeps.exclude` array. No need for an integration test — the bug surfaces only in real Vite setups that pre-bundle; the unit test pins the surface our config hook exposes.

## Out of scope

- Integration-contributed virtual modules: `INTEGRATION_SIDE_EFFECTS_VIRTUAL_ID` is the per-integration aggregate virtual. Individual integration virtuals (e.g., `virtual:swatchbook/tailwind.css`) are registered by integration factories and route through the integration's own virtual-module contract — those don't need addon-level exclusion.

--- a/.changeset/optimize-deps-exclude-virtuals.md
+++ b/.changeset/optimize-deps-exclude-virtuals.md
@@ -1,0 +1,10 @@
+---
+'@unpunnyfuns/swatchbook-core': patch
+'@unpunnyfuns/swatchbook-addon': patch
+'@unpunnyfuns/swatchbook-blocks': patch
+'@unpunnyfuns/swatchbook-switcher': patch
+'@unpunnyfuns/swatchbook-integrations': patch
+'@unpunnyfuns/swatchbook-mcp': patch
+---
+
+The addon's Vite plugin now excludes its virtual module IDs (`virtual:swatchbook/tokens`, `virtual:swatchbook/integration-side-effects`) from Vite's `optimizeDeps` pre-bundling. Vite uses esbuild for pre-bundling, and esbuild doesn't see Rollup-style `resolveId` hooks — a preview file that gets pulled into pre-bundling and imports one of our virtuals would fail with `Could not resolve "virtual:swatchbook/tokens"`. The exclusion routes the virtuals through the dev-time plugin pipeline where `resolveId` / `load` handle them. Build mode is unaffected (Rollup throughout).

--- a/packages/addon/src/virtual/plugin.ts
+++ b/packages/addon/src/virtual/plugin.ts
@@ -70,6 +70,28 @@ export function swatchbookTokensPlugin({
     name: 'swatchbook:virtual-tokens',
     enforce: 'pre',
 
+    /**
+     * Vite uses esbuild for `optimizeDeps` pre-bundling (development
+     * mode), and esbuild doesn't see Rollup-style `resolveId` hooks.
+     * Without this exclusion, a preview file that imports
+     * `virtual:swatchbook/tokens` (directly or via the addon's
+     * `useToken` hook) can get pulled into pre-bundling, hitting
+     * esbuild with no resolver registered and failing with
+     * `Could not resolve "virtual:swatchbook/tokens"`.
+     *
+     * Excluding the virtual IDs tells Vite to route them through the
+     * Rollup-style pipeline at request time — our `resolveId` / `load`
+     * hooks above handle the resolution. Build mode is unaffected
+     * (build uses Rollup throughout).
+     */
+    config() {
+      return {
+        optimizeDeps: {
+          exclude: [VIRTUAL_MODULE_ID, INTEGRATION_SIDE_EFFECTS_VIRTUAL_ID],
+        },
+      };
+    },
+
     async buildStart() {
       // Skip the redundant load when preset.viteFinal already supplied
       // a freshly-loaded project via `initialProject`. The first HMR

--- a/packages/addon/test/virtual-plugin.test.ts
+++ b/packages/addon/test/virtual-plugin.test.ts
@@ -6,7 +6,7 @@
 // of the cost.
 import type { Config, Project } from '@unpunnyfuns/swatchbook-core';
 import { expect, it } from 'vitest';
-import { collectWatchPaths } from '#/virtual/plugin.ts';
+import { collectWatchPaths, swatchbookTokensPlugin } from '#/virtual/plugin.ts';
 
 const CWD = '/project';
 
@@ -64,4 +64,26 @@ it('handles absolute paths untouched', () => {
   const config: Config = { tokens: ['/abs/tokens/**/*.json'] };
   const paths = collectWatchPaths(config, undefined, CWD);
   expect(paths).toEqual(['/abs/tokens']);
+});
+
+it('excludes the addon virtual IDs from Vite optimizeDeps pre-bundling', () => {
+  // Vite uses esbuild for optimizeDeps, which doesn't see Rollup-style
+  // `resolveId` hooks. The config hook tells Vite to route the virtuals
+  // through the dev-time plugin pipeline instead of esbuild's bundler.
+  const plugin = swatchbookTokensPlugin({
+    config: { tokens: ['tokens/**/*.json'] },
+    cwd: '/project',
+  });
+  const configResult =
+    typeof plugin.config === 'function'
+      ? plugin.config({}, { command: 'serve', mode: 'development' })
+      : undefined;
+  const exclude =
+    configResult && 'optimizeDeps' in configResult ? configResult.optimizeDeps?.exclude : undefined;
+  expect(exclude).toEqual(
+    expect.arrayContaining([
+      'virtual:swatchbook/tokens',
+      'virtual:swatchbook/integration-side-effects',
+    ]),
+  );
 });

